### PR TITLE
Add IBuddyList

### PR DESCRIPTION
**Pull request by MidoriKami**
Something of note is I added `[MemberNotNullWhen(...)]` attributes to the boolean checks, this enables some very nice syntactical sugar for the dev.

image

### CODE REVIEW

File: Dalamud/Plugin/Services/IBuddyList.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+/// It does not include the local player.
+/// </summary>
+public interface IBuddyList : IReadOnlyCollection<BuddyMember>
+{
+    /// <summary>
+    /// Gets the amount of battle buddies the local player has.
+    /// </summary>
+    public int Length { get; }
+    
+    /// <summary>
```
Comment (KazWolfe) [inline Dalamud/Plugin/Services/IBuddyList.cs:19]: Unsure if either of these should be part of the public API, given that they're just checking the relevant `Buddy` for `null` already.

File: Dalamud/Plugin/Services/IBuddyList.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    /// </summary>
+    [MemberNotNullWhen(true, "CompanionBuddy")]
+    public bool CompanionBuddyPresent { get; }
+    
+    /// <summary>
+    /// Gets a value indicating whether the local player's pet is present.
+    /// </summary>
+    [MemberNotNullWhen(true, "PetBuddy")]
+    public bool PetBuddyPresent { get; }
+
```
Comment (KazWolfe) [inline Dalamud/Plugin/Services/IBuddyList.cs:30]: Unsure if either of these should be part of the public API, given that they're just checking the relevant `Buddy` for null already.

Comment (MidoriKami) [reply Dalamud/Plugin/Services/IBuddyList.cs:30]: That's fair, in normal use I would more likely just check for null like this:
image

Comment (KazWolfe) [reply Dalamud/Plugin/Services/IBuddyList.cs:30]: Hm. Unless there's a way for these objects to be not-null but still not be "present" (previously spawned but no longer around?), let's remove these from the interface and mark them deprecated in the base service.

File: Dalamud/Game/ClientState/Buddy/BuddyList.cs [hunk 3]
```diff
@@ -16,7 +17,10 @@ namespace Dalamud.Game.ClientState.Buddy;
 [PluginInterface]
 [InterfaceVersion("1.0")]
 [ServiceManager.BlockingEarlyLoadedService]
-public sealed partial class BuddyList : IServiceType
+#pragma warning disable SA1015
```
Comment (karashiiro) [inline Dalamud/Game/ClientState/Buddy/BuddyList.cs:20]: This is fixed in https://github.com/DotNetAnalyzers/StyleCopAnalyzers/issues/3641, we should pull in the latest prerelease version of the analyzer. Not the responsibility of this PR, though.

**Review by avafloww — APPROVED** [review]

**Pull request merged**